crates/matrix-sdk/src/deduplicating_handler.rs

```diff
@@ -55,31 +68,67 @@ impl<Key: Clone + Ord + std::hash::Hash> DeduplicatingHandler<Key> {
     ) -> Result<()> {
         let mut map = self.inflight.lock().await;
 
-        if let Some(mutex) = map.get(&key).cloned() {
+        if let Some(request_mutex) = map.get(&key).cloned() {
             // If a request is already going on, await the release of the lock.
             drop(map);
 
-            return mutex.lock().await.map_err(|()| Error::ConcurrentRequestFailed);
+            let mut request_guard = request_mutex.lock().await;
+
+            return match *request_guard {
+                QueryState::Success => {
+                    // The query completed with a success: forward this success.
+                    Ok(())
+                }
+
+                QueryState::Failure => {
+                    // The query completed with an error, but we don't know what it is; report
+                    // there was an error.
+                    Err(Error::ConcurrentRequestFailed)
+                }
+
+                QueryState::NotFinishedYet => {
+                    // If we could take a hold onto the mutex without it being in the success or
+                    // failure state, then the query hasn't completed (e.g. it could have been
+                    // cancelled). Repeat it.
+                    //
+                    // Note: there might be other waiters for the deduplicated result; they will
+                    // still be waiting for the mutex above, since the mutex is obtained for at
+                    // most one holder at the same time.
+                    self.run_code(key, code, &mut request_guard).await
+                }
+            };
         }
 
-        // Assume a successful request; we'll modify the result in case of failures
-        // later.
-        let request_mutex = Arc::new(Mutex::new(Ok(())));
+        // Start at the `None` state to indicate we haven't completed the request yet.
```

Comment:
That's not the `None` state :sweat_smile:. I see now why you picked `NotFinishedYet`, though it still seems that we could say something like:

> Let's assume the cancelled state, if we succeed or fail we'll modify the result.

Reply:
Haha, remnant from a previous push. Thanks, I've used Cancelled in lieu of NotFinishedYet, and tweaked the comment accordingly.

(Also added a note about the `code` future which must be idempotent if run multiple times.)